### PR TITLE
feat: prune oversaturated bins when depth increases

### PR DESCRIPTION
**Pull request by istae**
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)

### CODE REVIEW

**Review by acud — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: 2 of 3 files reviewed, 4 unresolved discussions (waiting on @istae, @metacertain, and @mrekucci)

___
*[pkg/topology/kademlia/kademlia.go, line 585 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIs_qWEZ2mkEUXP4bn:-MhIs_qX9dc_jWgou4p7:btw95bm) ([raw file](https://github.com/ethersphere/bee/blob/2278e599ccba664eac084a8a8c8384f709fa1107/pkg/topology/kademlia/kademlia.go#L585)):*
> ```Go
> 
> 		if lenPeers > saturationPeers {
> 			peersToRemove, err := randomSubset(peers, lenPeers-saturationPeers)
> ```

This is up for debate, but we might like to favor long-lasting peers here. you could utilize the kademlia peer metrics and find the `n` most long-connected peers and favor them, kicking out the rest. This has been a tactic used successfully on other DHTs
___
*[pkg/topology/kademlia/kademlia.go, line 587 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIswA1BCGeFaS3SYKk:-MhIswA1BCGeFaS3SYKl:b-4okfvz) ([raw file](https://github.com/ethersphere/bee/blob/2278e599ccba664eac084a8a8c8384f709fa1107/pkg/topology/kademlia/kademlia.go#L587)):*
> ```Go
> 			peersToRemove, err := randomSubset(peers, lenPeers-saturationPeers)
> 			if err != nil {
> 				continue
> ```

if the error is ignored, why return it?
___
*[pkg/topology/kademlia/kademlia.go, line 1159 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIzOmk7n1IDbm8zqDT:-MhIzOmk7n1IDbm8zqDU:b5c1h9g) ([raw file](https://github.com/ethersphere/bee/blob/69f460c8db5932f72c938889a055049b5b56c575/pkg/topology/kademlia/kademlia.go#L1159)):*
> ```Go
> }
> 
> // IsBalanced returns if Kademlia is balanced to bin.
> ```

why is this removed?
also, when removing peers from a bin which is oversaturated, keeping bin balancing in mind will help us save some future operations. it would be nice to just drop the peers which are the newest and do not provide us balancing on the bin at the highest priority
___
*[pkg/topology/kademlia/kademlia_test.go, line 307 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIzf0D3aZJuQ2KBOoL:-MhIzf0D3aZJuQ2KBOoM:b-5m3vkt) ([raw file](https://github.com/ethersphere/bee/blob/69f460c8db5932f72c938889a055049b5b56c575/pkg/topology/kademlia/kademlia_test.go#L307)):*
> ```Go
> }
> 
> func TestManageWithBalancing(t *testing.T) {
> ```

not sure why is this removed

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -570,6 +574,25 @@ func (k *Kad) manage() {
...
 
+func (k *Kad) pruneOversaturatedBins(depth uint8) {
+
+	for i := uint8(0); i < depth; i++ {
+
+		peers := k.connectedPeers.BinPeers(i)
+		lenPeers := len(peers)
+
+		if lenPeers > saturationPeers {
+			peersToRemove, err := randomSubset(peers, lenPeers-saturationPeers)
```
Comment (metacertain) [inline pkg/topology/kademlia/kademlia.go:585]: use overSaturationPeers instead of saturationPeers

File: pkg/topology/kademlia/kademlia.go [hunk 2]
```diff
@@ -570,6 +574,25 @@ func (k *Kad) manage() {
...
 }
 
+func (k *Kad) pruneOversaturatedBins(depth uint8) {
+
+	for i := uint8(0); i < depth; i++ {
+
+		peers := k.connectedPeers.BinPeers(i)
+		lenPeers := len(peers)
+
+		if lenPeers > saturationPeers {
```
Comment (metacertain) [inline pkg/topology/kademlia/kademlia.go:584]: use overSaturationPeers instead of saturationPeers so its only pruned if oversaturation limit is breached

**Review by metacertain — CHANGES_REQUESTED** [review]

**Review by istae — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: 2 of 3 files reviewed, 7 unresolved discussions (waiting on @acud, @istae, and @mrekucci)

___
*[pkg/topology/kademlia/kademlia.go, line 585 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIs_qWEZ2mkEUXP4bn:-MhNLY7_BDKR2e0VJbri:b-6k72cx) ([raw file](https://github.com/ethersphere/bee/blob/2278e599ccba664eac084a8a8c8384f709fa1107/pkg/topology/kademlia/kademlia.go#L585)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

This is up for debate, but we might like to favor long-lasting peers here. you could utilize the kademlia peer metrics and find the `n` most long-connected peers and favor them, kicking out the rest. This has been a tactic used successfully on other DHTs
</blockquote></details>

This is actually a great suggestion, thanks.
___
*[pkg/topology/kademlia/kademlia.go, line 1159 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhNL8x-3SE16cExiaew:-MhNL8x-3SE16cExiaex:ba2w2ed) ([raw file](https://github.com/ethersphere/bee/blob/69f460c8db5932f72c938889a055049b5b56c575/pkg/topology/kademlia/kademlia.go#L1159)):*
> ```Go
> }
> 
> // IsBalanced returns if Kademlia is balanced to bin.
> ```

`IsBalanced` is not used anywhere externally and the unit test was affected by my changes so I felt there is no use maintaining it.

**Review by mrekucci — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: all files reviewed, 9 unresolved discussions (waiting on @acud and @istae)

___
*[pkg/topology/kademlia/kademlia.go, line 584 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW2VCU1BwS1XNn59G-:-MiW2VCU1BwS1XNn59G0:bo6kglt) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia.go#L584)):*
> ```Go
> // pruneOversaturatedBins disconnects peers from out of depth, oversaturated bins
> // while maintaining the balance of the bin and favoring peers with longers connections
> func (k *Kad) PruneOversaturatedBins(depth uint8) {
> ```

`pruneOversaturatedBins` -> `PruneOversaturatedBins`; I'd also rephrase the "from out of depth" part, probably something like "with depth bigger then the given param"...
___
*[pkg/topology/kademlia/kademlia.go, line 594 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW4DeU1zaVQG5pXFdU:-MiW4DeU1zaVQG5pXFdV:bsx74p7) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia.go#L594)):*
> ```Go
> 		binPeers := k.connectedPeers.BinPeers(uint8(i))
> 		binPeersCount := len(binPeers)
> 		if len(binPeers) < overSaturationPeers {
> ```

You can use `binPeersCount` instead of `len(binPeers)`, since you already have it.
___
*[pkg/topology/kademlia/kademlia.go, line 605 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW4xr19x-dVhNM4Oi_:-MiW4xr19x-dVhNM4Oia:b4sz5cc) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia.go#L605)):*
> ```Go
> 			peers := k.balancedSlotPeers(pseudoAddr, binPeers, i)
> 
> 			if len(peers) > 1 {
> ```

You can do early `continue` here to save some indentation (with `if len(peers) == 0 { continue }`).
___
*[pkg/topology/kademlia/kademlia.go, line 615 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW5xq95-vbrFFqJtqD:-MiW5xqA2mAF4JEoJkgO:b-5ioevi) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia.go#L615)):*
> ```Go
> 					}
> 				}
> 				_ = k.p2p.Disconnect(newestPeer)
> ```

I'd recommend at least log the error.
___
*[pkg/topology/kademlia/kademlia_test.go, line 1178 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW72ogEAE9QDrframe:-MiW72ogEAE9QDrframf:bx6hzpj) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia_test.go#L1178)):*
> ```Go
> }
> 
> func TestOutofDepthPrune(t *testing.T) {
> ```

Would be great if you could document the sleep constants: why or for what we're waiting for.

**Review by istae — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: all files reviewed, 7 unresolved discussions (waiting on @acud and @metacertain)

___
*[pkg/topology/kademlia/kademlia.go, line 584 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhEEnUkP8y6AGY13Mbc-r1-584:-MiWHWZd6b7mt61xHbqH:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/2278e599ccba664eac084a8a8c8384f709fa1107/pkg/topology/kademlia/kademlia.go#L584)):*
<details><summary><i>Previously, metacertain wrote…</i></summary><blockquote>

use overSaturationPeers instead of saturationPeers so its only pruned if oversaturation limit is breached
</blockquote></details>

Done.
___
*[pkg/topology/kademlia/kademlia.go, line 585 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhEEnUkP8y6AGY13Mbc-r1-585:-MiWHXa-3hjhYCOCkod7:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/2278e599ccba664eac084a8a8c8384f709fa1107/pkg/topology/kademlia/kademlia.go#L585)):*
<details><summary><i>Previously, metacertain wrote…</i></summary><blockquote>

use overSaturationPeers instead of saturationPeers
</blockquote></details>

Done.
___
*[pkg/topology/kademlia/kademlia.go, line 587 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIswA1BCGeFaS3SYKk:-MiWHdQr3Yo5JUO-4ibz:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/2278e599ccba664eac084a8a8c8384f709fa1107/pkg/topology/kademlia/kademlia.go#L587)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

if the error is ignored, why return it?
</blockquote></details>

Done.
___
*[pkg/topology/kademlia/kademlia.go, line 1159 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIzOmk7n1IDbm8zqDT:-MiWICMV5n65ezQPg3He:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/69f460c8db5932f72c938889a055049b5b56c575/pkg/topology/kademlia/kademlia.go#L1159)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

why is this removed?
also, when removing peers from a bin which is oversaturated, keeping bin balancing in mind will help us save some future operations. it would be nice to just drop the peers which are the newest and do not provide us balancing on the bin at the highest priority
</blockquote></details>

Done.
___
*[pkg/topology/kademlia/kademlia_test.go, line 307 at r1](https://reviewable.io/reviews/ethersphere/bee/2412#-MhIzf0D3aZJuQ2KBOoL:-MiWIDR5B8tIujv0-nhC:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/69f460c8db5932f72c938889a055049b5b56c575/pkg/topology/kademlia/kademlia_test.go#L307)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

not sure why is this removed
</blockquote></details>

Done.

**Review by acud — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r2, 1 of 1 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: 3 of 5 files reviewed, 4 unresolved discussions (waiting on @acud, @istae, @metacertain, and @mrekucci)

___
*[pkg/topology/kademlia/kademlia.go, line 584 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW2VCU1BwS1XNn59G-:-MiWhzbW2GdCXJXjj1av:b1lnqfj) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia.go#L584)):*
<details><summary><i>Previously, mrekucci wrote…</i></summary><blockquote>

`pruneOversaturatedBins` -> `PruneOversaturatedBins`; I'd also rephrase the "from out of depth" part, probably something like "with depth bigger then the given param"...
</blockquote></details>

this doesn't necessarily needs to be exported. it could be exported through `export_test` just for the sake of expose it to the test
___
*[pkg/topology/kademlia/kademlia_test.go, line 1233 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiWjatu-XIyTdMCdfhq:-MiWjatvD99H08pM3N9f:b-u27ed1) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia_test.go#L1233)):*
> ```Go
> 
> 	// check that bin 0 is balanced
> 	waitBalanced(t, kad, uint8(0))
> ```

as previously mentioned, it would be nice to have a test that checks that there were at least _some_ unbalanced bins. this can be done by adding peers with the `Connected` hook, instead of adding a peer and expecting it to be dialed (it probably won't in unit test environment). Using `Connected` can ensure you can reach an oversaturated state on every bin you wish, thereby ensuring the the pruning works more concisely

**Review by istae — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: 3 of 6 files reviewed, 4 unresolved discussions (waiting on @acud, @metacertain, and @mrekucci)

___
*[pkg/topology/kademlia/kademlia.go, line 584 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiW2VCU1BwS1XNn59G-:-Mi_eRYdENhqPjdDc_bP:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia.go#L584)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

this doesn't necessarily needs to be exported. it could be exported through `export_test` just for the sake of expose it to the test
</blockquote></details>

Done.
___
*[pkg/topology/kademlia/kademlia_test.go, line 1233 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiWjatu-XIyTdMCdfhq:-Mi_eT9t3VF7M19MX254:boi8cw1) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia_test.go#L1233)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

as previously mentioned, it would be nice to have a test that checks that there were at least _some_ unbalanced bins. this can be done by adding peers with the `Connected` hook, instead of adding a peer and expecting it to be dialed (it probably won't in unit test environment). Using `Connected` can ensure you can reach an oversaturated state on every bin you wish, thereby ensuring the the pruning works more concisely
</blockquote></details>

rand.Seed(2) at the beginning of the test, as it turns out, assures bin 0 and 1 are balanced, but not 2. I also added an explicit check in the test for this.

**Review by acud — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)* status: 3 of 6 files reviewed, 4 unresolved discussions (waiting on @acud, @istae, @metacertain, and @mrekucci)

___
*[pkg/topology/kademlia/kademlia_test.go, line 1233 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiWjatu-XIyTdMCdfhq:-Miu5E0b7wziZLxHFOpg:b1tw3wd) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia_test.go#L1233)):*
<details><summary><i>Previously, istae (istae) wrote…</i></summary><blockquote>

rand.Seed(2) at the beginning of the test, as it turns out, assures bin 0 and 1 are balanced, but not 2. I also added an explicit check in the test for this.
</blockquote></details>

that's a bit fragile and any change cannot be made to the test as it would break the expected "randomness". why not go with a predefined scenario?

**Comment by istae** [comment]
> > _[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2412)_ status: 3 of 6 files reviewed, 4 unresolved discussions (waiting on @acud, @istae, @metacertain, and @mrekucci)
> 
> _[pkg/topology/kademlia/kademlia_test.go, line 1233 at r6](https://reviewable.io/reviews/ethersphere/bee/2412#-MiWjatu-XIyTdMCdfhq:-Miu5E0b7wziZLxHFOpg:b1tw3wd) ([raw file](https://github.com/ethersphere/bee/blob/61d35c5dbbb4173bb69bc4dd4984bd5d273cb2f6/pkg/topology/kademlia/kademlia_test.go#L1233)):_
> 
> Previously, istae (istae) wrote…
> > rand.Seed(2) at the beginning of the test, as it turns out, assures bin 0 and 1 are balanced, but not 2. I also added an explicit check in the test for this.
> 
> that's a bit fragile and any change cannot be made to the test as it would break the expected "randomness". why not go with a predefined scenario?

This was a quick way to generate some bins that meet the requirements, the other solution requires precise construction of addresses to balance bins. I will work on achieving this.

**Review by acud — APPROVED** [review]
Nicely done!

**Pull request merged**